crepecake/src/main/java/com/google/cloud/tools/crepecake/builder/BuildConfiguration.java

```diff
@@ -0,0 +1,177 @@
+    TARGET_IMAGE_NAME,
+    /** The image tag of the built image (the part after the colon). */
+    TARGET_TAG,
+
+    /** The credential helper name used by {@link DockerCredentialRetriever}. */
+    CREDENTIAL_HELPER_NAME,
+  }
+
+  public static class Builder {
+
+    /** Textual descriptions of the configuration fields. */
+    @VisibleForTesting
+    static final Map<Fields, String> FIELD_DESCRIPTIONS =
+        new EnumMap<Fields, String>(Fields.class) {
+          {
+            put(Fields.BASE_IMAGE_SERVER_URL, "base image registry server URL");
+            put(Fields.BASE_IMAGE_NAME, "base image name");
+            put(Fields.TARGET_SERVER_URL, "target registry server URL");
+            put(Fields.TARGET_IMAGE_NAME, "target image name");
+            put(Fields.TARGET_TAG, "target image tag");
+            put(Fields.CREDENTIAL_HELPER_NAME, "credential helper name");
+          }
+        };
+
+    private static final String MISSING_FIELD_MESSAGE_SUFFIX =
+        " required but not set in build configuration";
+
+    private Map<Fields, String> values = new EnumMap<>(Fields.class);
+
+    private Builder() {}
+
+    public Builder setBaseImageServerUrl(String baseImageServerUrl) {
+      values.put(Fields.BASE_IMAGE_SERVER_URL, baseImageServerUrl);
+      return this;
+    }
+
+    public Builder setBaseImageName(String baseImageName) {
+      values.put(Fields.BASE_IMAGE_NAME, baseImageName);
+      return this;
+    }
+
+    public Builder setTargetServerUrl(String targetServerUrl) {
+      values.put(Fields.TARGET_SERVER_URL, targetServerUrl);
+      return this;
+    }
+
+    public Builder setTargetImageName(String targetImageName) {
+      values.put(Fields.TARGET_IMAGE_NAME, targetImageName);
+      return this;
+    }
+
+    public Builder setTargetTag(String targetTag) {
+      values.put(Fields.TARGET_TAG, targetTag);
+      return this;
+    }
+
+    public Builder setCredentialHelperName(String credentialHelperName) {
+      values.put(Fields.CREDENTIAL_HELPER_NAME, credentialHelperName);
+      return this;
+    }
+
+    /**
+     * @return the corresponding build configuration
+     * @throws IllegalStateException if required fields were not set
+     */
+    public BuildConfiguration build() {
+      List<String> descriptions = new ArrayList<>();
+      for (Fields field : Fields.values()) {
+        if (!values.containsKey(field)) {
+          descriptions.add(FIELD_DESCRIPTIONS.get(field));
+        }
+      }
+      switch (descriptions.size()) {
+        case 0:
+          values = Collections.unmodifiableMap(values);
+          return new BuildConfiguration(values);
+
+        case 1:
+          throw new IllegalStateException(descriptions.get(0) + MISSING_FIELD_MESSAGE_SUFFIX);
+
+        case 2:
+          throw new IllegalStateException(
+              descriptions.get(0) + " and " + descriptions.get(1) + MISSING_FIELD_MESSAGE_SUFFIX);
```

Comment:
I feel like there should be a utility method for this formatting pattern somewhere but I can't find it right now.

Reply:
Added a TODO to find one.